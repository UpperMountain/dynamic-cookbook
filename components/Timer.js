import React from "react";
import { Text, StyleSheet } from "react-native";

const styles = StyleSheet.create({
  clock: {
    fontSize: 36,
    lineHeight: 36
  }
});

class Timer extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      intervalId: undefined,
      time: this.props.seconds
    };
  }

  tick = () => {
    if (this.props.active) {
      this.setState({ time: this.state.time - 1 });
    }
  };

  componentDidMount() {
    this.setState({ intervalId: setInterval(this.tick, 1000) });
  }

  componentWillUnmount() {
    clearInterval(this.state.intervalId);
  }

  componentDidUpdate() {
    if (this.state.time <= 0) {
      clearInterval(this.state.intervalId);
      this.props.onFinish ? this.props.onFinish() : null;
    }
  }

  pad = n => {
    return n < 10 ? "0" + n : n;
  };

  formatTime = preTimer => {
    let out = "";
    if (this.props.seconds >= 3600) {
      let hrs = Math.floor(this.state.time / 3600) % 60;
      if (preTimer) {
        out += hrs + "h ";
      } else {
        out += this.pad(hrs) + ":";
      }
    }
    if (this.props.seconds >= 60) {
      let mins = Math.floor(this.state.time / 60) % 60;
      if (preTimer) {
        if (mins > 0) {
          out += mins + "m ";
        }
      } else {
        out += this.pad(mins) + ":";
      }
    }
    let secs = this.state.time % 60;
    if (preTimer) {
      if (secs > 0) {
        out += secs + "s";
      }
    } else {
      out += this.pad(secs);
    }

    return out;
  };

  render() {
    return (
      <Text style={styles.clock}>{this.formatTime(!this.props.active)}</Text>
    );
  }
}

export default Timer;
