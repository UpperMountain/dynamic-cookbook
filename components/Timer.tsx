import React from "react";
import { Text, StyleSheet } from "react-native";
import { OnGoingTimer } from "../lib/dependencyTree";

const styles = StyleSheet.create({
  clock: {
    fontSize: 24,
    lineHeight: 24
  }
});

interface Props {
  onTick: (id: number) => void;
  id: number;
  timer: OnGoingTimer;
  active: boolean;
  onFinish: () => void;
}

interface State {
  intervalId: NodeJS.Timeout | undefined;
}

class Timer extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      intervalId: undefined
    };
  }

  tick = () => {
    if (this.props.active) {
      this.props.onTick(this.props.id);
    }
  };

  componentDidMount() {
    this.setState({ intervalId: setInterval(this.tick, 1000) });
  }

  componentWillUnmount() {
    clearInterval(this.state.intervalId as NodeJS.Timeout);
  }

  componentDidUpdate() {
    if (this.props.timer.elapsed >= this.props.timer.duration) {
      clearInterval(this.state.intervalId as NodeJS.Timeout);
      this.props.onFinish ? this.props.onFinish() : null;
    }
  }

  pad = (n: number) => {
    return n < 10 ? "0" + n : n;
  };

  formatTime = (preTimer: boolean) => {
    let left = this.props.timer.duration - this.props.timer.elapsed;
    let out = "";
    if (this.props.timer.duration >= 3600) {
      let hrs = Math.floor(left / 3600) % 60;
      if (preTimer) {
        out += hrs + "h ";
      } else {
        out += this.pad(hrs) + ":";
      }
    }
    if (this.props.timer.duration >= 60) {
      let mins = Math.floor(left / 60) % 60;
      if (preTimer) {
        if (mins > 0) {
          out += mins + "m ";
        }
      } else {
        out += this.pad(mins) + ":";
      }
    }
    let secs = left % 60;
    if (preTimer) {
      if (secs > 0) {
        out += secs + "s";
      }
    } else {
      out += this.pad(secs);
    }

    return out;
  };

  render() {
    return (
      <Text style={styles.clock}>{this.formatTime(!this.props.active)}</Text>
    );
  }
}

export default Timer;
