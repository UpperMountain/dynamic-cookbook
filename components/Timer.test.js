import React from "react";
import renderer from "react-test-renderer";
import Timer from "./Timer";

describe("Timer", () => {
  it("should render a timer without exploding", () => {
    renderer.create(<Timer seconds={10} />);
  });

  it("should render a timer which desplays 2h 12m 8s", () => {
    const cmp = renderer.create(<Timer seconds={7928} />);
    expect(cmp.toJSON()).toMatchSnapshot();
  });
});
