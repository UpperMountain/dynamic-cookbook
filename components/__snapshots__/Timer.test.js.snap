// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer should render a timer which desplays 2h 12m 8s 1`] = `
<Text
  style={
    Object {
      "fontSize": 36,
      "lineHeight": 36,
    }
  }
>
  2h 12m 8s
</Text>
`;
